### ISSUE TITLE
Data-modification operations for graphs

### ISSUE THREAD

**Issue by mknittig** [arangodb/arangodb#1202]
AFAIK ArangoDB currently supports only data-modification operations on documents. I would like to have data-modification operations on graphs too (like Cypher in neo4j). 

My use case: I have a relationship like this:
catogory1-[:subcategory]->category2

It would be great if I could extend it to the structure with one AQL command (creating the new documents and edges):
catogory1-[:subcategory]->category2-[:subcategory]->catogory3-[:subcategory]->category4

**Comment by fceller** [comment]
Are you referring to the general-graph module? Or to AQL?

AQL supports modifications via https://docs.arangodb.com/Aql/DataModification.html This can also be used on edges and vertices. But not in command.

general-graph: Maybe it is possible touse the Fluent interface (https://docs.arangodb.com/General-Graphs/FluentAQLInterface.html) and some map operations or extend the interface slightly.

**Comment by mknittig** [comment]
I was referring to AQL. Can you give ma an example how to do that? Haven't seen this on my search for such a functionality..

**Comment by jsteemann** [comment]
Do I understand correctly that the task is to 
* create two vertices (i.e. one in category3 and one in category4) 
* create the connecting subcategory edges between these vertices (i.e. from category2 to category3 and from category3 to category4)
* accomplish this in one go?

I think this won't work in a single AQL query, because it would modify data in multiple collections at the same time, and AQL is currently limited to modifying a single collection per query.

It is however possible to split the one task into multiple commands to achieve the same result. Here is an example (assumes two collections are used and they are named `categories` and `subcategories`):
```js
db._executeTransaction({
  collections: {
    write: [ "categories", "subcategories" ]
  },
  action: function (p) {
    /* insert categories */
    var cat1 = db.categories.insert(p.category1Data);
    var cat2 = db.categories.insert(p.category2Data);
    var cat3 = db.categories.insert(p.category3Data);
    var cat4 = db.categories.insert(p.category4Data);
    /* insert subcategories, connecting the categories */
    db.subcategories.insert(cat1, cat2, p.subcategory12Data);
    db.subcategories.insert(cat2, cat3, p.subcategory23Data);
    db.subcategories.insert(cat3, cat4, p.subcategory34Data);
  },
  params: {
    category1Data: ...,
    category2Data: ...,
    category3Data: ...,
    category4Data: ...,
    subcategory12Data: ...,
    subcategory23Data: ...,
    subcategory34Data: ...
  }
});
```
There are probably other ways to achieve the same result (e.g. by issuing multiple AQL queries, using the fluent interface), but the above should work already and is very generic, so it can be applied in other situations, too.

**Comment by mknittig** [comment]
Yes, that's exactly what I want. Cypher (neo4j) supports this (even with FOREACH).
I know that there are workarounds, but I think it would be a useful feature.

**Comment by fceller** [comment]
@mknittig Could you give us an example, how to write this in Cypher?

**Comment by mknittig** [comment]
http://neo4j.com/docs/stable/cypherdoc-create-nodes-and-relationships.html
Really like the cypher syntax for graphs..

**Comment by dothebart** [comment]
Hi, 
You can now do this with ArangoDBs new traversal (introduced with 2.8): https://docs.arangodb.com/3.0/AQL/Graphs/Traversals.html

We also have a new general introducion to graphs in ArangoDB: https://docs.arangodb.com/3.0/Manual/Graphs/index.html

The last missing feature has been implemented with 3.0, you now also can alter `_from` and `_to`. 
We discussed with users about their queries on stackoverflow:

http://stackoverflow.com/search?q=[arangodb]+graph

ArangoDB 3.0 RC2 is available for download through: https://www.arangodb.com/download-preview/
